### PR TITLE
initial commit for adding scope attribute

### PR DESCRIPTION
**Pull request by jennyf19**
a bit more work todo & clean-up.
have tested w/
- [x] azure functions
- [x] gRPC
- [x] web app calls web api
<table>
<tr><td>VerifyUserHasAnyAcceptedScope</td><td>RequiredScopeAttribute</td></tr>
<tr><td>

```csharp
 static readonly string[] scopeRequiredByApi = new string[] { "access_as_user" };

// GET: api/values
[HttpGet]
public async Task<IEnumerable<Todo>> GetAsync()
{
    HttpContext.VerifyUserHasAnyAcceptedScope(scopeRequiredByApi);
...
```
</td><td>

```csharp
[Authorize]
[Route("api/[controller]")]
[RequiredScope("access_as_user")] 
public class TodoListController : Controller
{...

[HttpGet]
public async Task<IEnumerable<Todo>> GetAsync()
{
    string owner = User.GetDisplayName();
...
```
</td></tr>
<tr><td>
When using incorrect scope: 

```
HttpRequestException: 403 Forbidden IDW10203: The 'scope' or 'scp' claim does not contain scopes 'access_as_cat' or was not found.
Microsoft.Identity.Web.DownstreamWebApiGenericExtensions.ConvertToOutput<TOutput>(HttpResponseMessage response) in DownstreamWebApiGenericExtensions.cs, line 273
```
</td><td>
When using incorrect scope: 

```
HttpRequestException: 403 Forbidden
Microsoft.Identity.Web.DownstreamWebApiGenericExtensions.ConvertToOutput<TOutput>(HttpResponseMessage response) in DownstreamWebApiGenericExtensions.cs, line 273
```
</td></tr>
<tr><td>
</td><td>
Updated: 

```
HttpRequestException: 403 Forbidden IDW10203: The 'scope' or 'scp' claim does not contain scopes 'access_as_cat' or was not found.
Microsoft.Identity.Web.DownstreamWebApiGenericExtensions.ConvertToOutput<TOutput>(HttpResponseMessage response) in DownstreamWebApiGenericExtensions.cs, line 273
```
</table>

### CODE REVIEW

File: src/Microsoft.Identity.Web/Resource/RequiredScopeAttribute.cs [hunk 1]
```diff
@@ -23,5 +33,13 @@ public RequiredScopeAttribute(params string[] acceptedScopes)
         {
             Arguments = new object[] { acceptedScopes };
         }
+
+        /// <summary>
+        ///
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web/Resource/RequiredScopeAttribute.cs:38]: /// <summary>
        /// Default constructor for the attribute. It's meant to be used with the RequiredScopesSettings property.
        /// </summary>
        /// <example><code>[RequiredScope(RequiredScopesSettings="AzureAd:Scopes")]</code></example>

File: src/Microsoft.Identity.Web/Resource/RequiredScopeFilter.cs [hunk 2]
```diff
@@ -41,7 +43,15 @@ public void OnAuthorization(AuthorizationFilterContext context)
...
-            else if (context.HttpContext.User == null || context.HttpContext.User.Claims == null || !context.HttpContext.User.Claims.Any())
+
+            GetEffectiveScopes(context);
+
+            ValidateEffectiveScopes(context);
+        }
+
+        internal void ValidateEffectiveScopes(AuthorizationFilterContext context)
+        {
+            if (context.HttpContext.User == null || context.HttpContext.User.Claims == null || !context.HttpContext.User.Claims.Any())
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web/Resource/RequiredScopeFilter.cs:54]: I think we'd want to add a test that `_effectiveAcceptedScopes` is not **null** (and throw if that's the case with an explicit exception telling the developer that they need provide either `RequiredScopesSettings` or `RequiredScopesSettings` in the `RequiredScope` attribute.

**Review by jmprieur — COMMENTED** [review]
LGTM. I think it's an awesome API now. It's exciting that it can be used at the level of the controller/page and the actions. 

We'd want to fix the static analysis issues (the suggestion about throwing when _effectiveAcceptedScopes is null, should help)

File: src/Microsoft.Identity.Web/Resource/RequiredScopeAttribute.cs (hunk 1 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/Resource/RequiredScopeAttribute.cs:38]: fixed. thank you.

---
In reply to: [573118532](https://github.com/AzureAD/microsoft-identity-web/pull/942#discussion_r573118532) [](ancestors = 573118532)

File: src/Microsoft.Identity.Web/Resource/RequiredScopeFilter.cs (hunk 2 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/Resource/RequiredScopeFilter.cs:54]: fixed. but let's discuss.

---
In reply to: [573122624](https://github.com/AzureAD/microsoft-identity-web/pull/942#discussion_r573122624) [](ancestors = 573122624)

**Review by jmprieur — COMMENTED** [review]
I had several suggestions, so I proposed a branch: https://github.com/AzureAD/microsoft-identity-web/compare/jennyf/scopeAttribute...jmprieur/scopeAttributeProposal?expand=1
